### PR TITLE
Refactor with Sync 1.0 changes

### PR DESCRIPTION
**Pull request by superkhau**
Connect to strongloop-internal/scrum-loopback#214

### CODE REVIEW

**Comment by bajtos** [comment]
Can you please make the commit message of 7a545e7021fdc69d0cc96cd803f6722b1ac3414e more explanatory? Change tracking was already enabled.

**Comment by superkhau** [comment]
@bajtos Running `grunt` alone shows this error:

```
Running "karma:unit" (karma) task
INFO [karma]: Karma v0.12.31 server started at http://localhost:8080/
INFO [launcher]: Starting browser PhantomJS
WARN [watcher]: Pattern "/Users/sh/work-repos/loopback-example-offline-sync/client/ngapp/test/mock/**/*.js" does not match any file.
INFO [PhantomJS 1.9.8 (Mac OS X)]: Connected on socket z-UTHdFtwZ2ZLg01gBEa with id 33976604
PhantomJS 1.9.8 (Mac OS X) ERROR
  TypeError: 'undefined' is not a function (evaluating 'Number.isFinite(parseInt(v, 10))')
  at .../loopback-example-offline-sync/client/lbclient/browser.bundle.js:16224


Warning: Task "karma:unit" failed. Use --force to continue.

Aborted due to warnings.
```

Do you know what this means?

**Comment by superkhau** [comment]
Also getting:

```shell
Running "mochaTest:server" (mochaTest) task

  Todo
    New todo empty data
      POST /api/Todos
        ✓ should be allowed
        ✓ should have statusCode 200
        1) should respond with a new todo # this is failing
  ...
  1 failing

  1) Todo New todo empty data POST /api/Todos should respond with a new todo:
     AssertionError: null == true
      at Context.<anonymous> (server/test/todo.test.js:23:9)
```

For some reason, the test is returning an actual mongo id (might be something to do with the "before save" operation hook).

File: server/test/todo.test.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ describe('Todo', function() {
       });
 
       it('should respond with a new todo', function() {
+        console.log(this.res.body.id);
```
Comment (superkhau) [inline server/test/todo.test.js]: `this.res.body.id` returns an actual MongoDB id instead of the `t-math` id.

File: server/test/todo.test.js [hunk 2]
```diff
@@ -19,7 +19,8 @@ describe('Todo', function() {
         assert.equal(this.res.statusCode, 200);
       });
 
-      it('should respond with a new todo', function() {
+      it.skip('should respond with a new todo', function() {
+        console.log(this.res.body.id);
```
Comment (superkhau) [inline server/test/todo.test.js]: `this.res.body.id` returns an actual MongoDB id instead of the `t-math` id. Any idea why this is happening?

**Comment by bajtos** [comment]
> Number.isFinite

See https://github.com/strongloop/loopback-boot/pull/124

> Todo New todo empty data POST /api/Todos should respond with a new todo

No idea.

> this.res.body.id returns an actual MongoDB id instead of the t-math id. Any idea why this is happening?

I think that's because all Todo ids are initialized with a GUID value - see `"defaultFn": "guid"` in model config. This is intended and required for sync to work correctly.

File: client/lbclient/boot/replication.js [hunk 3]
```diff
@@ -23,17 +23,33 @@ module.exports = function(client) {
...
-    if (client.network.isConnected) {
-      RemoteTodo.replicate(LocalTodo, function() {
-        LocalTodo.replicate(RemoteTodo, cb);
+  var since = {push: -1, pull: -1};
+  function sync() {
+    LocalTodo.replicate(
+      RemoteTodo,
+      since.push,
+      function pushed(err, conflicts, cps) {
+        console.log('LR', arguments);
```
Comment (superkhau) [inline client/lbclient/boot/replication.js]: Getting "[Error: Invalid URI "/api/Todos/checkpoint", undefined, undefined]" here also.

File: common/models/todo.js [hunk 4]
```diff
@@ -5,10 +5,11 @@ module.exports = function(Todo) {
 
   Todo.definition.properties.created.default = Date.now;
 
-  Todo.beforeSave = function(next, model) {
-    if (!model.id) model.id = 't-' + Math.floor(Math.random() * 10000).toString();
+  Todo.observe('before save', function(ctx, next) {
+    if (!ctx.Model.id)
+      ctx.Model.id = 't-' + Math.floor(Math.random() * 10000).toString();
```
Comment (bajtos) [inline common/models/todo.js]: This hook is no longer needed, the id property is initialised by loopback to a new GUID.

I think the idea here was to use small semi-unique numbers that are easier to read than full GUIDs.

I am proposing to make a following change in the GUI to make IDs easier to read:
 - take the first 8 characters of the generated uid (the 4 low bytes of the timestamp)
 - use [proquint-](https://www.npmjs.com/package/proquint-) to convert these 8 characters into a readable string
 - show this readable representation in all place where we were rendering the original id (mostly conflict resolution views?)

Feel free to make UI changes in a new pull request. However, the hook should be removed as part of this patch.

File: common/models/user.json [hunk 5]
```diff
@@ -1,11 +1,14 @@
...
   "base": "PersistedModel",
+  "strict": "throw",
+  "persisteUndefinedAsNull": true,
   "trackChanges": true,
   "properties": {
     "id": {
       "id": true,
-      "type": "string"
+      "type": "string",
+      "defaultFn": "guid"
```
Comment (bajtos) [inline common/models/user.json]: The User model does not support replication/sync ATM. Please revert these changes.

File: client/lbclient/boot/replication.js (hunk 3 above)
Comment (bajtos) [inline client/lbclient/boot/replication.js]: Is the URL "/api/Todos/checkpoint" available in the explorer?

Please create a new RemoteTodo model per the instructions in [docs](http://docs.strongloop.com/display/LB/Synchronization#Synchronization-Createaclientapp) and use it instead of the base Todo model here.

```diff
- var RemoteTodo = client.models.Todo;
+ var RemoteTodo = client.models.RemoteTodo;
```

Comment (bajtos) [inline client/lbclient/boot/replication.js]: I think the problem is caused by the fact that your RemoteTodo (=base Todo ATM) has change tracking enabled, thus it's trying to rectify changes via REST API. That's why you need to create a new subclass RemoteTodo which will not track changes but still provide the replication API. See the docs for more details.

> For each replicated model, create two new client-only subclasses:
 - A local model that will use local storage to persist the changes offline
 - A remote model that will be connected to the server and used as a target for replication. This model will have change tracking disabled (because the server is already handling it) and enable only the replication REST API.

File: common/models/user.json (hunk 5 above)
Comment (superkhau) [inline common/models/user.json]: Should I also remove `"trackChanges": true` (this was here before my changes)?

File: client/lbclient/boot/replication.js [hunk 6]
```diff
@@ -23,17 +23,34 @@ module.exports = function(client) {
...
-      RemoteTodo.replicate(LocalTodo, function() {
-        LocalTodo.replicate(RemoteTodo, cb);
+  var since = {push: -1, pull: -1};
+  function sync() {
+    console.log('sync triggered');
+    LocalTodo.replicate(
+      RemoteTodo,
+      since.push,
+      function pushed(err, conflicts, cps) {
+        if (err) throw err;
```
Comment (superkhau) [inline client/lbclient/boot/replication.js]: Still getting `Invalid URI "/api/Todos/checkpoint"` here.

File: common/models/todo.js (hunk 4 above)
Comment (superkhau) [inline common/models/todo.js]: Hook removed.

File: common/models/user.json [hunk 7]
```diff
@@ -1,7 +1,6 @@
 {
   "name": "user",
   "base": "PersistedModel",
-  "trackChanges": true,
```
Comment (superkhau) [inline common/models/user.json]: Should we remove this also? It was already there before my changes.

**Comment by superkhau** [comment]
>See https://github.com/strongloop/loopback-boot/pull/124

I see, will bump `loopback-boot` version when that gets merged.

>No idea.

Going to skip the test then.

>I think that's because all Todo ids are initialized with a GUID value - see `"defaultFn": "guid"` in model config. This is intended and required for sync to work correctly.

I see. So it's possible this initialization is causing that one test to fail then. Anyways, I'm skipping the test, so its not blocking what we're trying to do anyways.

File: client/lbclient/datasources.local.js [hunk 8]
```diff
@@ -2,6 +2,6 @@ var GLOBAL_CONFIG = require('../../global-config');
 
 module.exports = {
   remote: {
-    url: GLOBAL_CONFIG.restApiUrl
+    url: 'http://localhost:3000' + GLOBAL_CONFIG.restApiUrl
```
Comment (superkhau) [inline client/lbclient/datasources.local.js]: I figured out the problem. This might be a bug, but you need to add the FULL URL for things to work properly. Thoughts?

File: client/lbclient/boot/replication.js (hunk 3 above)
Comment (superkhau) [inline client/lbclient/boot/replication.js]: Figured it out, see https://github.com/strongloop/loopback-example-offline-sync/pull/60/files#r28929070

File: client/lbclient/datasources.local.js (hunk 8 above)
Comment (bajtos) [inline client/lbclient/datasources.local.js]: See https://github.com/strongloop/strong-remoting/issues/200

The host + port must not be hard-coded here. Check the way how GLOBAL_CONFIG is constructed and modify the code to produce a full URL, so that `datasources.local.js` can stay unchanged.

Comment (bajtos) [inline client/lbclient/datasources.local.js]: Note that host + port value must be synchronized with settings in `server/config.json`. When you change port in `server/config.json`, the client app should automatically pick it up after rebuild via grunt.

File: common/models/user.json (hunk 5 above)
Comment (bajtos) [inline common/models/user.json]: Yes please.

File: common/models/user.json (hunk 7 above)
Comment (bajtos) [inline common/models/user.json]: Yes plase.

**Comment by bajtos** [comment]
> Going to skip the test then.

No no, don't skip the test, fix it instead. As you can see, it's checking the format of the id property:

```js
assert(this.res.body.id.match(/t-\d+/));
```

Since you are no longer generating ids as `t-XYZA`, you should rework the test to check something else that makes sense. For example that id has a GUID format.

Or simply remove `it('should respond with a new todo')` completely, since the other tests are already testing that a new Todo instance can be created.

**Comment by superkhau** [comment]
>No no, don't skip the test...

Updated the test to check for proper GUID format.

File: client/ngapp/scripts/controllers/todo.js [hunk 9]
```diff
@@ -137,23 +146,18 @@ angular.module('loopbackExampleFullStackApp')
...
-    }
+    conflict.resolveUsingTarget(refreshConflicts);
   };
 
   $scope.resolveManually = function(conflict) {
-    conflict.manual.save(function(err) {
-      if(err) return errorCallback(err);
-      conflict.resolve(refreshConflicts);
-    });
+    conflict.resolveManually({
```
Comment (superkhau) [inline client/ngapp/scripts/controllers/todo.js]: I figured this out by digging into your source code, but we definitely need some actual docs explaining how to resolve conflicts manually.

**Comment by superkhau** [comment]
@bajtos Please code review. If you see anything obvious that needs fixing, please add it to this PR yourself as I will be away for 2 weeks. If there is no hurry to merge this, then I can finish it up when I'm back. As for the easy-to-read GUID feature, I created a separate issue at https://github.com/strongloop/loopback-example-offline-sync/issues/62 and will work on it after I'm back.

File: client/ngapp/scripts/controllers/todo.js (hunk 9 above)
Comment (bajtos) [inline client/ngapp/scripts/controllers/todo.js]: Let's improve the docs then! See https://github.com/strongloop/loopback/issues/1333

File: client/lbclient/datasources.json [hunk 10]
```diff
@@ -1,6 +1,7 @@
 {
   "remote": {
-    "connector": "remote"
+    "connector": "remote",
+    "url": "/api"
```
Comment (bajtos) [inline client/lbclient/datasources.json]: AFAIK, this value is overriden by [datasources.local.js](https://github.com/strongloop/loopback-example-offline-sync/blob/feature/refactor-with-sync-1.0-changes/client/lbclient/datasources.local.js) and thus not used.

File: client/ngapp/scripts/controllers/todo.js [hunk 11]
```diff
@@ -56,9 +62,11 @@ angular.module('loopbackExampleFullStackApp')
...
 
   $scope.addTodo = function () {
-    var todo = new Todo({title: $scope.newTodo});
-    todo.save();
-    $scope.newTodo = '';
+    Todo.create({title: $scope.newTodo})
+      .then(function() {
+        $scope.newTodo = '';
+        $scope.$apply();
+      });
```
Comment (bajtos) [inline client/ngapp/scripts/controllers/todo.js]: @ritch do you happen to remember why your original code was cleaning `$scope.newTodo` before the `save` request finished? Is it important to preserve that behaviour, or is the new solution proposed by @superkhau ok too?

File: client/ngapp/scripts/controllers/todo.js [hunk 12]
```diff
@@ -137,23 +145,18 @@ angular.module('loopbackExampleFullStackApp')
...
 
   $scope.resolveManually = function(conflict) {
-    conflict.manual.save(function(err) {
-      if(err) return errorCallback(err);
-      conflict.resolve(refreshConflicts);
-    });
+    conflict.resolveManually({
+      completed: conflict.manual.completed,
+      title: conflict.manual.title
+    }, refreshConflicts);
```
Comment (bajtos) [inline client/ngapp/scripts/controllers/todo.js]: I believe this can be simplified to `conflict.resolveManually(conflict.manual, refreshConflicts)`

File: global-config.js [hunk 13]
```diff
@@ -10,4 +12,10 @@ exports.restApiRoot = '/api';
...
 // Replace with a full url (including hostname) if your client is being
 // served from a different server than your REST API.
-exports.restApiUrl = exports.restApiRoot;
+exports.restApiUrl = url.format({
+  protocol: 'http',
+  slashes: true,
+  hostname: serverConfig.host,
+  port: serverConfig.port,
+  pathname: 'api'
+});
```
Comment (bajtos) [inline global-config.js]: Sorry if I have mislead you. The server config depends on global-config (see [server/config.local.js](https://github.com/strongloop/loopback-example-offline-sync/blob/master/server/config.local.js)), not the other way around.

If we need host and port config, then it should be set here in global-config and then used wherever needed.

**Comment by chandadharap** [comment]
Assigning back to @bajtos , Simon on PTO

**Comment by bajtos** [comment]
Reassigning back to Simon, offline sync is not urgent now.

File: client/lbclient/datasources.json (hunk 10 above)
Comment (superkhau) [inline client/lbclient/datasources.json]: Removed `url` k/v.

File: client/ngapp/scripts/controllers/todo.js (hunk 11 above)
Comment (superkhau) [inline client/ngapp/scripts/controllers/todo.js]: @ritch ^

Comment (ritch) [inline client/ngapp/scripts/controllers/todo.js]: ya this is fine

**Comment by ritch** [comment]
@superkhau @bajtos is this good to go?

**Comment by superkhau** [comment]
@ritch Almost. I need to refactor the configs based on @bajtos's comments and then it should be good to go.

File: client/ngapp/scripts/controllers/todo.js (hunk 12 above)
Comment (superkhau) [inline client/ngapp/scripts/controllers/todo.js]: Simplified. ;)

**Comment by superkhau** [comment]
@bajtos Code review again please.

File: client/lbclient/boot/replication.js [hunk 14]
```diff
@@ -23,17 +23,32 @@ module.exports = function(client) {
   };
 
   // setup model replication
+  var since = {push: -1, pull: -1};
```
Comment (bajtos) [inline client/lbclient/boot/replication.js]: Nitpick: AFAIK, we use Node coding style where curly braces are surrounded by spaces: `{ push: -1, pull: -1 }`.

**Comment by bajtos** [comment]
Nice! I left one more comment, feel free to ignore it. The rest LGTM.

File: client/lbclient/boot/replication.js (hunk 14 above)
Comment (superkhau) [inline client/lbclient/boot/replication.js]: Is there a style guide for the style you describe? I've just been following the Google one at https://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml

**Comment by superkhau** [comment]
>Nitpick: AFAIK, we use Node coding style where curly braces are surrounded by spaces: { push: -1, pull: -1 }.

Is there a style guide for the style you describe? I've just been following the Google Javascript style guide at https://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml

File: client/lbclient/boot/replication.js (hunk 14 above)
Comment (bajtos) [inline client/lbclient/boot/replication.js]: > Is there a style guide for the style you describe?

https://strongloop.atlassian.net/wiki/display/SL/Code+Style+Guide

We have had some discussions about spacing in array/object initialisation - see https://strongloop.atlassian.net/wiki/display/SL/Code+Style+Guide?focusedCommentId=21430339#comment-21430339

Comment (superkhau) [inline client/lbclient/boot/replication.js]: TY, I will take a look. ;)
